### ISSUE TITLE
Phone validation too strict

### ISSUE THREAD

**Issue by matthew-dean** [formvalidation/support#72]
The phone validation reports that "(555) 866-8404" is an invalid US phone number, even though that's how it would be most commonly written. It validates "(555)866-8404" (without a space), which is an uncommon way to write it.

**Comment by nghuuphuoc** [comment]
@matthew-dean 
Thank for the feedback. I will improve it.
Currently, you can use the [transformer](http://formvalidation.io/settings/#validator-transformer) option.

```javascript
$(yourForm).formValidation({
    fields: {
        phoneNumber: {
            validators: {
                phone: {
                    country: 'US',
                    transformer: function($field, validatorName, validator) {
                        // Get the field value
                        var value = $field.val();

                        // Remove all space
                        value = value.replace(/\s/g, '');

                        // Return new value
                        return value;
                    }
                }
            }
        }
    }
});
```

When user enter ```(555) 866-8404```, it will be treated as ```(555)866-8404```, and then becomes a valid US phone number.

More example:
[Modifying the value before validating](http://formvalidation.io/examples/modifying-value-before-validating/)

**Comment by matthew-dean** [comment]
I was thinking a little bit more about it. Since punctuation in phone numbers if just a way to organize digits, really a phone number should be considered "valid" if it has the right number of numerical values.

For example, these should all be considered valid US phone numbers:

* 555.866.8404
* 555 • 866 • 8404
* 555~866~8404

They're unconventionally written, but they're still valid *numbers*. What would be great is to simply validate them as numbers, and then if I wanted to store / display them in a standard way, to apply post-validation formatting, to convert to something more standard like:

* (555) 866-8404

So, in the case of phone numbers (and some other validators), the validator really shouldn't care about white space or punctuation. If it's recognizable and unambiguous as input, it should pass. (IMHO)

**Comment by matthew-dean** [comment]
Other common ways to write phone numbers:

* 555.866.8404 ext. 1200
* 555-866-8404 x1200

In those cases, only the first piece of it needs to be validated / formatted, and can stop validating once it reaches an alpha character (or could optionally validate that the extension is a valid number? and could return the phone and extension as two separate pieces of data?). However, I know that's a more complex scenario, just wanted to put that out there.

**Comment by nghuuphuoc** [comment]
I've added new [Validating international phone numbers](http://formvalidation.io/examples/validating-international-phone-numbers/) example that might be useful to you.
Behinde the scences, it uses the Google libphonenumber patterns, so it covers a lot of possible formats.

**Comment by mrtristan** [comment]
@matthew-dean has an example that is seemingly not covered by your post. He's talking about US phone numbers that have an extension at the end, not so much another country.

i think this is a bug because other countries offer this ability. see the United Kingdom examples that you offer:

+44161234567x123
+44 (0) 161 1234567 ext 123

this does not work for the US.

**Comment by matthew-dean** [comment]
Right, a phone number field should not prevent entry (or should have an option not to). It should validate the first part of the field that's recognizable as a phone number for a location and leave the rest intact. See Microsoft Outlook for a good example of this. It formats whatever you put in the phone number field of a contact, but you can put whatever you want into it, and it will just format the part that it thinks is a phone number.
